src/Microsoft.ML.Core/Data/ModelHeader.cs

```diff
@@ -351,10 +351,19 @@ public static void CheckVersionInfo(ref ModelHeader header, VersionInfo ver)
             Contracts.CheckDecode(header.ModelSignature == ver.ModelSignature, "Unknown file type");
             Contracts.CheckDecode(header.ModelVerReadable <= header.ModelVerWritten, "Corrupt file header");
             if (header.ModelVerReadable > ver.VerWrittenCur)
-                throw Contracts.ExceptDecode("Cause: ML.NET {0} cannont read component '{1}' of the model, because the model is too new.\n" +
+                throw Contracts.ExceptDecode("Cause: ML.NET {0} cannot read component '{1}' of the model, because the model is too new.\n" +
                                 "Suggestion: Make sure the model is trained with ML.NET {0} or older.\n" +
                                 "Debug details: Maximum expected version {2}, got {3}.",
                                 typeof(VersionInfo).Assembly.GetName().Version, ver.LoaderSignature, header.ModelVerReadable, ver.VerWrittenCur);
+            var envVar = Environment.GetEnvironmentVariable("NEWER_MODEL_VERSION_ALLOWED");
+            if (envVar != "TRUE" && header.ModelVerWritten > ver.VerWrittenCur)
```

Comment:
Is `header.ModelVerWritten` & `ver.VerWrittenCur` stored once per model, or is it stored per component in the model?

Most time forward-breaking changes to components won't affect most models. Generally a single model users only a small number of components.

If it's not per-component, could we make it so? That way it will only throw when the version of the specifically used components are older than required by the model.

Reply:
In that case, then the component would be the one throwing the exception, like text-loader. Should I just try searching for other components that may also have this issue? If so, then there's still the possibility that the newer component may introduce silent bugs that we can't detect because the headers didn't change.

Reply:
It's an interesting trade-off.

#### Background - versions for training vs serving
My concern is the version people run in production (to serve the models) tends to be difficult to update. The version used on a user's local machine is easy to update. The impact of this is the models will be trained on a newer version of ML․NET causing it to throw this new version error and not be runnable. 

Two things will then occur, (1) the local machine will keep an old version of ML․NET pinned for model training, or (2) a rather bit of hassle updating the serving side to match the trained models.

Generally, forward-breaking changes are rare (besides when introducing new components). Generally the version numbers in the components are pretty good at catching forward-breaking changes. I don't recall examples of _silent_ forward-breaking changes.

##### Per-component checks
Is there a way to catch the deserialization errors for the components, then check for and append the version error? 

Example of the deserialization error: ([via](https://github.com/dotnet/machinelearning/issues/5303#issuecomment-657040153))
```c#
(1) Unexpected exception: Error during class instantiation, 'System.InvalidOperationException'
   at Microsoft.ML.Runtime.ComponentCatalog.LoadableClassInfo.CreateInstanceCore(Object[] ctorArgs) in /Users/justinormont/Documents/src/machinelearning/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:line 254
   at Microsoft.ML.Runtime.ComponentCatalog.TryCreateInstance[TRes](IHostEnvironment env, Type signatureType, TRes& result, String name, String options, Object[] extra) in /Users/justinormont/Documents/src/machinelearning/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:line 1028
   at Microsoft.ML.Runtime.ComponentCatalog.TryCreateInstance[TRes,TSig](IHostEnvironment env, TRes& result, String name, String options, Object[] extra) in /Users/justinormont/Documents/src/machinelearning/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:line 993
```

Perhaps we can always warn where the runner is older than the model, but augment the component deserialization error to make it more actionable?

By appending the version error, we incase the usability the error message to be actionable by the user. Currently, it's very generic and leaves a user without the next steps. The error message you put in is really good as it says, the model version and serving version, and next steps for the user to correct the error.

---

I do really like that we're addressing the class of version errors here. Fixing things twice, [once for the user](https://github.com/dotnet/machinelearning/issues/5303#issuecomment-657040153), and then fixing the class of errors which fixes it for future users.

Reply:
Hi @justinormont. If I understand correctly, you suggest that I switch this general error message into a warning (note that the reason why I made it an error was because warnings get logged and might be missed by the user), and throw per-component error.